### PR TITLE
Resumable Workers

### PR DESCRIPTION
**Pull request by fabeschan**

### CODE REVIEW

**Comment by fabeschan** [comment]
Yeah. there's probably a security vulnerability where a worker can spoof a dead one.

Perhaps validate `remote` field and make sure it matches? Although that can be spoofed too

**Comment by fabeschan** [comment]
Given the fact that a `resume_bundle` request still checks whether the particular worker is authorized to handle the bundle's owner's bundles, the possibility of malicious spoofing is limited to workers of the same owner. So I don't think this is a big issue

**Comment by kashizui** [comment]
Yeah, owners of workers have arbitrary control for what they do to actually execute the bundle and return the contents, so there's no point in adding extra security for that.

As long as no one can spoof the dead worker of another user (or more precisely, spoof the worker for a bundle that they have no write access to) we should already meet our guarantees

**Comment by fabeschan** [comment]
@kashizui @percyliang review pls

File: codalab/lib/download_manager.py
Comment (kashizui) [inline codalab/lib/download_manager.py:83]: 😍

File: codalab/rest/workers.py
Comment (kashizui) [inline codalab/rest/workers.py:137]: clean up print statements

File: codalab/common.py
Comment (kashizui) [inline codalab/common.py:104]: What's the rationale for splitting a KILLED state out from FAILED?

File: worker/run.py
Comment (kashizui) [inline worker/run.py:122]: ?

File: worker/run.py
Comment (kashizui) [inline worker/run.py:130]: try to use `logger.info` or `logger.debug` as appropriate instead of print statements

File: codalab/common.py
Comment (kashizui) [inline codalab/common.py:105]: Maybe we should have a state diagram showing the possible transitions between all these states..

**Review by kashizui — CHANGES_REQUESTED** [review]
Generally looks good to me! Excited for robust workers.

Probably a big ask but is there any way to test all of this?

**Comment by fabeschan** [comment]
@kashizui It's a perfectly reasonable ask to have tests covering this... I've been thinking about how to test this properly too, just don't have a good idea on how to go about doing it yet

Suggestions/brainstorm welcome!

**Comment by kashizui** [comment]
Small tests:

- simple units tests for saving and loading runs

Medium tests:

- ?

**Comment by kashizui** [comment]
it's tempting to try to come up with big integration tests to test how we think the worker should behave but those tests are so hard to maintain..

File: codalab/model/bundle_model.py
Comment (percyliang) [inline codalab/model/bundle_model.py:693]: extra `#`

File: codalab/model/worker_model.py
Comment (percyliang) [inline codalab/model/worker_model.py:328]: If we actually expect this to fail, shouldn't we have exponential backoff?

**Review by percyliang — COMMENTED** [review]
This is great!

Have you tested this code with shared_file_system?

File: codalab/common.py
Comment (fabeschan) [reply codalab/common.py:104]: To me, it makes sense to differentiate between a user aborted run and a run that failed and dies on its own. Too many times I've panicked before realizing a run shown as 'failed' was actually something I killed a while ago.

Perhaps the bigger reason is any signal that can give clues as to whether a failure is the user's fault or potentially our platform's fault is valuable in a practical sense.

File: codalab/model/worker_model.py
Comment (fabeschan) [reply codalab/model/worker_model.py:328]: In principle, yes, expo backoff makes sense. Perhaps it will be more appropriate to switch all requests of this type to retry with expo back off in a separate PR tho.

**Comment by fabeschan** [comment]
@percyliang I have not tried with shared_file_system. How does one usually set up a shared file system and test it?

File: codalab/model/worker_model.py
Comment (percyliang) [reply codalab/model/worker_model.py:328]: Let's put a comment for now then.

File: codalab/common.py
Comment (percyliang) [reply codalab/common.py:104]: I think having `killed` is fine, but the thing is that I think we've implicitly assumed that we had a closed set of states, and some parts of the code assume that (also the public stats worksheet).  So we should check those places, and generally have a plan for thinking about states - are we going to keep on increasing the number of states, or is this it?

**Comment by percyliang** [comment]
The NLP machines use `shared_file_system`.  It's just a matter of putting a flag in the `config.json` file and passing the `--shared-file-system` flag to workers.

**Comment by kashizui** [comment]
Yeah for local testing just run both server and workers with the shared file system flag on the same machine, and it should just work!

**Comment by fabeschan** [comment]
Add note: will be unblocked once a fix for #778 is approved and merged

**Comment by fabeschan** [comment]
Closed for reboot: https://github.com/codalab/codalab-cli/pull/805

**Pull request closed without merging**
